### PR TITLE
Add basic functionality for Metroid Fusion

### PR DESCRIPTION
**Pull request by JeffGainsNGames**
Includes:

- Database (trickless logic has had multiple passes)
- All room images
- Starter Preset
- Basic GUI and Export/PDF components from new game generation command

### CODE REVIEW

**Comment by henriquegemignani** [comment]
Would be nice to compress down the map images. Right now it's 18MB which is quite a lot.
Running [pngquant](https://pngquant.org/) with just this `pngquant.exe --force --ext .png *` decreased it to 9MB.

If you could rebase the branch such that the original sized images aren't in history it'd be great (though we can also squash merge this PR).

**Comment by Miepee** [comment]
> Running [pngquant](https://pngquant.org/) with just this `pngquant.exe --force --ext .png *` decreased it to 9MB.

You can also add a few more arguments, like: `pngquant.exe --force --ext .png --strip --nofs --speed 1 --posterize 2 --quality 0-30 *` to reduce it further down to 6.9MB

Also please remove images of maps that are not used in the DB (i.e. the debug/unused main deck rooms, different layers of the same room etc.)

**Comment by JeffGainsNGames** [comment]
> Would be nice to compress down the map images. Right now it's 18MB which is quite a lot. Running [pngquant](https://pngquant.org/) with just this `pngquant.exe --force --ext .png *` decreased it to 9MB.
> 
> If you could rebase the branch such that the original sized images aren't in history it'd be great (though we can also squash merge this PR).

Done. New cover is 120KB, and maps are down to ~6.2MB after deleting unused/dupes and compressing.

**Comment by JeffGainsNGames** [comment]
the maps were all added in the initial commit so I have no idea how to fix that history myself..

**Comment by duncathan** [comment]
no big deal. there's gonna be a billion commits in this pr so we were probably gonna squash anyway

**Comment by JeffGainsNGames** [comment]
so checks pass, but are there any tests I need to add or anything that I can be doing to keep this moving on my end ahead of reviews?
I do understand the scale of this PR so not implying a rush.

File: randovania/games/fusion/logic_database/header.json [hunk 1]
```diff
@@ -0,0 +1,1811 @@
...
+                "extra": {}
+            },
+            "ShinesparkTrick": {
+                "long_name": "Shinespark Tricks",
+                "description": "The player may be expected to use shinesparking to traverse a room (beyond what's expected in vanilla).",
+                "extra": {}
+            },
+            "Knowledge": {
+                "long_name": "Knowledge",
+                "description": "The player may be expected to know alternative enemy/block weaknesses.",
```
Comment (Miepee) [inline randovania/games/fusion/logic_database/header.json:366]: Should proibably be rewritten

File: randovania/games/fusion/logic_database/header.json [hunk 2]
```diff
@@ -0,0 +1,1811 @@
...
+                "extra": {}
+            },
+            "HeatRuns": {
+                "long_name": "Damage Runs",
+                "description": "The player may be expected to traverse a room while taking damage from the environment (Heat/Lava/Cold/Electric)",
+                "extra": {}
+            },
+            "WallJump": {
+                "long_name": "Wall Jump",
+                "description": "The player may be expected to wall jump to traverse rooms vertically.",
```
Comment (Miepee) [inline randovania/games/fusion/logic_database/header.json:376]: Should add vanila expectation

File: randovania/games/fusion/logic_database/header.json [hunk 3]
```diff
@@ -0,0 +1,1811 @@
...
+    "layers": [
+        "default"
+    ],
+    "starting_location": {
+        "region": "Main Deck",
+        "area": "Docking Bay Hangar",
+        "node": "Ship"
+    },
+    "initial_states": {},
+    "minimal_logic": null,
```
Comment (Miepee) [inline randovania/games/fusion/logic_database/header.json:1534]: todo?

File: randovania/games/fusion/game_data.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            "What patcher does Randovania use?",
+            "Randovania supports the open source patcher MARS (Metroid Advance Randomizer System).",
+        ),
+        (
+            "Which versions of Fusion are supported?",
+            "Only the USA version of Fusion is supported with no current plans to support additional versions.",
+        ),
+        (
+            "I saved in a place I can't get out of, am I softlocked?",
+            "You can use the warp to start function in the pause menu by pressing L and confirming. "
```
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/game_data.py:76]: need to escape quotes whoops

**Comment by Miepee** [comment]
Can you also please add TODOs for all the basic/empty classes that need to be filled out later? Cosmetic options, configuration, patches etc.

File: randovania/games/fusion/logic_database/header.json [hunk 5]
```diff
@@ -0,0 +1,1811 @@
...
+            "MissileLauncher": {
+                "long_name": "Missile Launcher",
+                "max_capacity": 1,
+                "extra": {
+                    "item": "Missiles"
+                }
+            },
+            "Missiles": {
+                "long_name": "Missiles",
+                "max_capacity": 250,
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/logic_database/header.json:50]: the patcher schema supports a certain maximum (at least the last time I checked)

File: randovania/games/fusion/logic_database/header.json [hunk 6]
```diff
@@ -0,0 +1,1811 @@
...
+            "Level3Locks": {
+                "long_name": "Level 3 Locks",
+                "extra": {}
+            },
+            "Level4Locks": {
+                "long_name": "Level 4 Locks",
+                "extra": {}
+            },
+            "Arachnus": {
+                "long_name": "Arachnus",
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/logic_database/header.json:232]: any way we could phrase this with a word before the bosses to group them all together? 
I don't exactly like "after `defeated arachnus`" but that's what I'm getting at.

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/logic_database/header.json:232]: `Boss Arachnus Defeated` ?

File: randovania/games/fusion/logic_database/header.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/logic_database/header.json:50]: iircc the patcher's maximum is way higher than what you ahve here

File: randovania/games/fusion/logic_database/header.json [hunk 7]
```diff
@@ -0,0 +1,1811 @@
...
+            },
+            "IceBeam": {
+                "long_name": "Ice Beam",
+                "max_capacity": 1,
+                "extra": {
+                    "item": "IceBeam"
+                }
+            },
+            "MissileLauncher": {
+                "long_name": "Missile Launcher",
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/logic_database/header.json:42]: will I have to change short name and all related files?

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/logic_database/header.json:42]: Technically you dont *have* to change short name. But would be nicer while you dont guarantee compatibility yet.

File: randovania/games/fusion/logic_database/header.json (hunk 3 above)
Comment (JeffGainsNGames) [reply randovania/games/fusion/logic_database/header.json:1534]: yes

File: randovania/games/fusion/logic_database/header.json [hunk 8]
```diff
@@ -358,22 +358,22 @@
...
             "HeatRuns": {
                 "long_name": "Damage Runs",
-                "description": "The player may be expected to traverse a room while taking damage from the environment (Heat/Lava/Cold/Electric)",
+                "description": "The player may be expected to traverse a room while taking damage from the environment (Heat/Lava/Cold/Electric).",
                 "extra": {}
             },
             "WallJump": {
                 "long_name": "Wall Jump",
-                "description": "The player may be expected to wall jump to traverse rooms vertically.",
+                "description": "The player may be expected to wall jump to traverse rooms vertically. All vanilla locations where walljumps are expected are considered trickless.",
```
Comment (Miepee) [inline randovania/games/fusion/logic_database/header.json:376]: so wall jump or walljump? be consistent here

File: randovania/games/fusion/assets/hash_words.json [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    "Unit",
+    "Vines",
+    "Zone",
+    "Grey",
+    "Blue",
+    "Green",
+    "Yellow",
+    "Red",
+    "Adam",
+    "Bob",
```
Comment (Miepee) [inline randovania/games/fusion/assets/hash_words.json:111]: Do we want bob on the hashlist? to my knowledge its a community given name and sounds a bit weird to have here.

File: randovania/games/fusion/exporter/game_exporter.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    from randovania.lib import status_update_lib
+
+
+@dataclasses.dataclass(frozen=True)
+class FusionGameExportParams(GameExportParams):
+    input_path: Path
+    output_path: Path
+
+
+class FusionGameExporter(GameExporter):
```
Comment (Miepee) [inline randovania/games/fusion/exporter/game_exporter.py:20]: missing todo?

File: randovania/games/fusion/game_data.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    development_state=game.DevelopmentState.EXPERIMENTAL,
+    presets=[
+        {"path": "starter_preset.rdvpreset"},
+    ],
+    faq=[
+        (
+            "What patcher does Randovania use?",
+            "Randovania supports the open source patcher MARS (Metroid Advance Randomizer System). "
+            "This is a new patcher built from scratch and is unrelated to other Fusion Randomizers. "
+            "Any differences between vanilla or any other Randomizers are covered under the differences tab.",
```
Comment (Miepee) [inline randovania/games/fusion/game_data.py:70]:
```suggestion
            "Any differences between vanilla are covered under the differences tab.",
```

File: randovania/games/fusion/gui/dialog/cosmetic_patches_dialog.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+from randovania.gui.dialog.base_cosmetic_patches_dialog import BaseCosmeticPatchesDialog
+from randovania.gui.generated.fusion_cosmetic_patches_dialog_ui import Ui_FusionCosmeticPatchesDialog
+
+if TYPE_CHECKING:
+    from PySide6 import QtWidgets
+
+    from randovania.layout.base.cosmetic_patches import BaseCosmeticPatches
+
+
+class FusionCosmeticPatchesDialog(BaseCosmeticPatchesDialog, Ui_FusionCosmeticPatchesDialog):
```
Comment (Miepee) [inline randovania/games/fusion/gui/dialog/cosmetic_patches_dialog.py:15]: todo?

File: randovania/games/fusion/gui/dialog/game_export_dialog.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+from randovania.games.game import RandovaniaGame
+from randovania.gui.dialog.game_export_dialog import GameExportDialog, spoiler_path_for
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from randovania.interface_common.options import PerGameOptions
+
+
+class FusionGameExportDialog(GameExportDialog):
```
Comment (Miepee) [inline randovania/games/fusion/gui/dialog/game_export_dialog.py:17]: todo?

File: randovania/games/fusion/logic_database/header.json [hunk 14]
```diff
@@ -0,0 +1,1811 @@
...
+                }
+            },
+            "IceBeam": {
+                "long_name": "Ice Beam",
+                "max_capacity": 1,
+                "extra": {
+                    "item": "IceBeam"
+                }
+            },
+            "MissileLauncher": {
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/logic_database/header.json:41]: done

**Review by Miepee — APPROVED** [review]
Looks fine for me.

File: test/cli/commands/test_export_game.py [hunk 15]
```diff
@@ -1,15 +1,23 @@
...
 from randovania.games.game import RandovaniaGame
 
+game_ignore_list = [
+    RandovaniaGame.FUSION,
+]
+
 
 def test_add_parser_arguments_for(game_enum: RandovaniaGame) -> None:
+    if game_enum in game_ignore_list:
+        pytest.skip(f"{game_enum} is ignored for this test.")
```
Comment (henriquegemignani) [inline test/cli/commands/test_export_game.py:19]: This is not necessary. All the test does is get export params. What's happening is that Fusion doesn't have the new thing I added to all games

**Review by henriquegemignani — CHANGES_REQUESTED** [review]

File: test/cli/commands/test_export_game.py (hunk 15 above)
Comment (JeffGainsNGames) [reply test/cli/commands/test_export_game.py:19]: Do I just have to align fusions exporter with the similar changes in https://github.com/randovania/randovania/pull/6102/files#diff-ce2e449c3b38198543a262f2088cc209f644e2a48c85ac7fa403f976352cd7ec

Comment (henriquegemignani) [reply test/cli/commands/test_export_game.py:19]: https://github.com/randovania/randovania/pull/6086/files actually

Comment (JeffGainsNGames) [reply test/cli/commands/test_export_game.py]: thanks! done.

**Review by henriquegemignani — APPROVED** [review]

**Pull request merged**